File: backend/src/main/java/com/coing/domain/coin/trade/service/TradeService.java
package com.coing.domain.coin.trade.service;

import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentLinkedQueue;

import org.springframework.http.HttpStatus;
import org.springframework.messaging.simp.SimpMessageSendingOperations;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import com.coing.domain.coin.trade.dto.TradeDto;
import com.coing.domain.coin.trade.entity.Trade;
import com.coing.global.exception.BusinessException;
import com.coing.util.MessageUtil;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class TradeService {

	private final SimpMessageSendingOperations simpMessageSendingOperations;

	private final Map<String, Queue<TradeDto>> tradeListCache = new ConcurrentHashMap<>();

	private final Map<String, Double> vwapCache = new ConcurrentHashMap<>();
	private final Map<String, Double> totalTradeVolumeCache = new ConcurrentHashMap<>();
	private final Map<String, Double> totalTradeNumberCache = new ConcurrentHashMap<>();
	private final Map<String, Double> prevTradePriceCache = new ConcurrentHashMap<>();
	private final MessageUtil messageUtil;

	private final Map<String, Long> lastSentTime = new ConcurrentHashMap<>();
	private static final long THROTTLE_INTERVAL_MS = 200;
	private static final int maxListSize = 20;

	public List<TradeDto> getTrades(String market) {
		Queue<TradeDto> queue = tradeListCache.get(market);

		if (queue == null || queue.isEmpty()) {
			throw new BusinessException(
				messageUtil.resolveMessage("trade.not.found"),
				HttpStatus.NOT_FOUND
			);
		}

		return queue.stream().toList();
	}

	public void updateTrade(Trade trade) {
		String market = trade.getCode();
		double price = trade.getTradePrice();
		double volume = trade.getTradeVolume();

		// 부가 지표 캐시 업데이트
		cacheValues(market, volume);

		double vwap = calculateVWAP(market, price, volume);
		double averageTradeSize = calculateAverageTradeSize(market);
		double tradeImpact = calculateTradeImpact(market, price);

		TradeDto dto = TradeDto.of(trade, vwap, averageTradeSize, tradeImpact);

		publish(dto);

		// 원자적으로 처리하여 race condition 방지
		tradeListCache.compute(market, (key, queue) -> {
			if (queue == null) {
				queue = new ConcurrentLinkedQueue<>();
			}
			queue.add(dto);

			while (queue.size() > maxListSize) {
				queue.poll();
			}

			return queue;
		});
	}

	public void publish(TradeDto dto) {
		String market = dto.code();
		long now = System.currentTimeMillis();
		long lastSent = lastSentTime.getOrDefault(market, 0L);

		if (now - lastSent >= THROTTLE_INTERVAL_MS) {
			simpMessageSendingOperations.convertAndSend("/sub/coin/trade/%s".formatted(market), dto);
			lastSentTime.put(market, now);
		}
	}

	@Scheduled(cron = "0 0 0 * * *")
	public void resetCaches() {
		vwapCache.clear();
		totalTradeVolumeCache.clear();
		totalTradeNumberCache.clear();
		prevTradePriceCache.clear();
	}

	private void cacheValues(String market, double volume) {
		vwapCache.putIfAbsent(market, 0.0);
		totalTradeVolumeCache.putIfAbsent(market, 0.0);
		totalTradeNumberCache.putIfAbsent(market, 0.0);
		prevTradePriceCache.putIfAbsent(market, 0.0);

		double totalTradeVolume = totalTradeVolumeCache.get(market);
		double totalTradeNumber = totalTradeNumberCache.get(market);

		totalTradeVolumeCache.put(market, totalTradeVolume + volume);
		totalTradeNumberCache.put(market, totalTradeNumber + 1);
	}

	private double calculateVWAP(String market, double price, double volume) {
		double prevVWAP = vwapCache.get(market);
		double totalTradeVolume = totalTradeVolumeCache.get(market);

		if (totalTradeVolume == 0)
			return 0;

		return ((prevVWAP * totalTradeVolume) + (price * volume)) / (totalTradeVolume + volume);
	}

	private double calculateAverageTradeSize(String market) {
		double totalTradeVolume = totalTradeVolumeCache.get(market);
		double totalTradeNumber = totalTradeNumberCache.get(market);
		return totalTradeNumber == 0 ? 0 : totalTradeVolume / totalTradeNumber;
	}

	private double calculateTradeImpact(String market, double price) {
		double prevTradePrice = prevTradePriceCache.get(market);
		prevTradePriceCache.put(market, price);
		return prevTradePrice == 0 ? 0 : (price - prevTradePrice) / prevTradePrice * 100;
	}
}

File: backend/src/main/java/com/coing/domain/coin/trade/service/TradeService.kt
package com.coing.domain.coin.trade.service

import com.coing.domain.coin.common.port.CoinDataHandler
import com.coing.domain.coin.common.port.EventPublisher
import com.coing.domain.coin.trade.dto.TradeDto
import com.coing.domain.coin.trade.entity.Trade
import com.coing.domain.notification.entity.TradeImpact
import com.coing.domain.notification.service.PushService
import com.coing.global.exception.BusinessException
import com.coing.util.MessageUtil
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import lombok.extern.slf4j.Slf4j
import org.springframework.http.HttpStatus
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue

@Service
@Slf4j
class TradeService(
    private val eventPublisher: EventPublisher<TradeDto>,
    private val messageUtil: MessageUtil,
    private val pushService: PushService,
    private val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
) : CoinDataHandler<Trade> {

    private val tradeListCache = ConcurrentHashMap<String, ConcurrentLinkedQueue<TradeDto>>()
    private val vwapCache = ConcurrentHashMap<String, Double>()
    private val totalTradeVolumeCache = ConcurrentHashMap<String, Double>()
    private val totalTradeNumberCache = ConcurrentHashMap<String, Double>()
    private val prevTradePriceCache = ConcurrentHashMap<String, Double>()
    private val lastSentTime = ConcurrentHashMap<String, Long>()
    private val lastPushSentTime = ConcurrentHashMap<String, Long>()
    private val pushMutexMap = ConcurrentHashMap<String, Mutex>()

    companion object {
        private const val THROTTLE_INTERVAL_MS = 200L
        private const val MAX_LIST_SIZE = 20
        private const val PUSH_THROTTLE_INTERVAL_MS = 60_000L
    }

    override fun update(data: Trade) {
        val market = data.code
        val price = data.tradePrice
        val volume = data.tradeVolume

        // 캐시 갱신
        updateCaches(market, volume)

        // 지표 계산
        val vwap = calculateVWAP(market, price, volume)
        val averageTradeSize = calculateAverageTradeSize(market)
        val tradeImpact = calculateTradeImpact(market, price)

        val dto = TradeDto.of(data, vwap, averageTradeSize, tradeImpact)

        updateTradeListCache(market, dto)
        publish(dto)
        pushMessage(dto)
    }

    private fun updateTradeListCache(market: String, dto: TradeDto) {
        tradeListCache.compute(market) { _, queue ->
            val newQueue = queue ?: ConcurrentLinkedQueue<TradeDto>()
            newQueue.add(dto)
            while (newQueue.size > MAX_LIST_SIZE) newQueue.poll()
            newQueue
        }
    }

    fun getTrades(market: String): List<TradeDto> {
        val queue = tradeListCache[market]
        if (queue.isNullOrEmpty()) {
            throw BusinessException(
                messageUtil.resolveMessage("trade.not.found"),
                HttpStatus.NOT_FOUND
            )
        }
        return queue.toList()
    }

    fun pushMessage(dto: TradeDto) {
        val market = dto.code
        val now = System.currentTimeMillis()
        val mutex = pushMutexMap.computeIfAbsent(market) { Mutex() }

        coroutineScope.launch {
            mutex.withLock {
                val lastSent = lastPushSentTime[market] ?: 0L
                if (now - lastSent < PUSH_THROTTLE_INTERVAL_MS) return@withLock

                val absImpact = kotlin.math.abs(dto.tradeImpact)
                val direction = if (dto.tradeImpact > 0) "HIGH" else "LOW"
                val messageKey = if (dto.tradeImpact > 0) "high.trade" else "low.trade"

                TradeImpact.entries
                    .filter { it != TradeImpact.NONE && absImpact >= it.threshold }
                    .forEach { impact ->
                        val topic = "$market-$direction-${impact.name}"
                        val body = String.format(messageUtil.resolveMessage(messageKey), impact.threshold)

                        pushService.sendAsync(
                            title = market,
                            body = body,
                            marketCode = market,
                            topic = topic
                        )
                    }

                lastPushSentTime[market] = now
            }
        }
    }

    fun publish(dto: TradeDto) {
        val market = dto.code
        val now = System.currentTimeMillis()
        val lastSent = lastSentTime[market] ?: 0L

        if (now - lastSent >= THROTTLE_INTERVAL_MS) {
            eventPublisher.publish(dto)
            lastSentTime[market] = now
        }
    }

    override fun fallbackUpdate(lastUpdate: String) {
        tradeListCache.forEach { (code, queue) ->
            val latest = queue.lastOrNull()
            if (latest != null) {
                latest.isFallback = true
                latest.lastUpdate = lastUpdate
                publish(latest)
            }
        }
    }

    @Scheduled(cron = "0 0 0 * * *")
    fun resetCaches() {
        vwapCache.clear()
        totalTradeVolumeCache.clear()
        totalTradeNumberCache.clear()
        prevTradePriceCache.clear()
    }

    private fun updateCaches(market: String, volume: Double) {
        vwapCache.putIfAbsent(market, 0.0)
        prevTradePriceCache.putIfAbsent(market, 0.0)
        totalTradeVolumeCache.merge(market, volume) { old, new -> old + new }
        totalTradeNumberCache.merge(market, 1.0) { old, new -> old + new }
    }

    /**
     * VWAP(Volume-Weighted Average Price) 체결가 가중 평균 가격 계산
     */
    private fun calculateVWAP(market: String, price: Double, volume: Double): Double {
        val newTotalVolume = totalTradeVolumeCache[market] ?: volume
        val previousVolume = newTotalVolume - volume
        val previousVWAP = if (previousVolume <= 0.0) price else (vwapCache[market] ?: price)

        // VWAP = (이전 VWAP*이전 거래량 + 현재 거래가격*현재 거래량) / (이전 거래량 + 현재 거래량)
        val newVWAP = ((previousVWAP * previousVolume) + (price * volume)) / newTotalVolume

        vwapCache[market] = newVWAP
        return newVWAP
    }

    /**
     * 평균 거래 크기 계산
     */
    private fun calculateAverageTradeSize(market: String): Double {
        val totalVolume = totalTradeVolumeCache[market] ?: 0.0
        val totalNumber = totalTradeNumberCache[market] ?: 0.0
        return if (totalNumber == 0.0) 0.0 else totalVolume / totalNumber
    }

    /**
     * 거래 임팩트 계산
     */
    private fun calculateTradeImpact(market: String, price: Double): Double {
        val prevPrice = prevTradePriceCache[market] ?: 0.0
        prevTradePriceCache[market] = price
        return if (prevPrice == 0.0) 0.0 else (price - prevPrice) / prevPrice * 100
    }
}
